p4-16/spec/P4-16-spec.adoc

```diff
@@ -9547,35 +9547,35 @@ The next table lists restrictions on what kinds of calls can be made
 from which places in a P4 program.  Calling a parser, control, or
 table means invoking its `apply()` method. Calling a value-set means using it
 in a select expression. The row for `extern` describes where extern method
-calls can be made from.
+or function calls can be made from.
 
 One way that an extern can be called from the top level of a parser or
 control is in an initializer expression for a declared variable,
 e.g. `bit<32> x = rand.get();`.
 
-[.center,cols="3,^2,^2,^2,^2,^2,^2",width=90%]
+[.center,cols="3,^2,^2,^2,^2,^2,^2,^2",width=90%]
 |===
-| This type 6+^| can be called at run time from this place in a P4 program
+| This type 7+^| can be called at run time from this place in a P4 program
 
-|             .>| parser state .>| control apply block .>| parser or top level .>| action .>| extern .>| function
+|             .>| top level .>| parser state .>| control apply block .>| top level within a parser or control .>| action .>| extern .>| function
 
-| package     | N/A       | N/A     | N/A       | N/A     | N/A    | N/A
+| package     | N/A 	      | N/A       | N/A     | N/A       | N/A     | N/A    | N/A
 
-| parser      | yes       |  no     | no        | no      | no     | no
+| parser      | no            | yes       |  no     | no        | no      | no     | no
 
-| control     |  no       | yes     | no        | no      | no     | no
+| control     | no            |  no       | yes     | no        | no      | no     | no
 
-| extern      | yes       | yes     | yes       | yes     | no     | no
+| extern      | yes           | yes       | yes     | yes       | yes     | no     | yes
```

Comment:
Suggest adding the following text before this table to elaborate on the "yes" in the "top level" column for the row "extern":

While the language syntactically allows extern function or method calls at the top level of a program, an implementation may restrict which extern functions or methods it allows in these places.  For example, an implementation might only support extern functions or method calls with the annotation `@pure`.  Any such calls are executed before the first packet is processed by the device.

(I am less sure about whether the last sentence is worth including.  It could open up a can of worms that I have not yet thought of.)

Reply:
This is a related question on top-level function/method calls.

I believe the only place where top-level function/method calls can happen is as constant initializers, e.g.,

```p4
const bit<32> c = f(...);
```

because the syntax only allows:

```plaintext
declaration
    : constantDeclaration
    | externDeclaration
    | actionDeclaration
    | parserDeclaration
    | typeDeclaration
    | controlDeclaration
    | instantiation
    | errorDeclaration
    | matchKindDeclaration
    | functionDeclaration
    ;
```

Then, the function/methods *must* produce local compile-time known values, since constant initializers should be local compile-time known.

... but is it? According to section 18.1, no sorts of function/method call is categorized as local compile-time known.

Reply:
Good catch.

It seems _possible_ that we could define a subset of function/method calls as local compile-time known values, but this seems like it might be a fairly nit-picky thing to do well, e.g. only pure functions whose parameters are all local compile-time known.  It seems very odd to me to call an expression with a function call, where the function call includes side effects on extern state, a local compile-time known expression, but perhaps it could be reasonable.  I just don't know if we want to go that far into the rabbit hole on this one.

Perhaps we just put "no" for function/method calls in the "top level" column of the table for now, and generalize it later if someone really wants to?

Are any function/method calls supported at the top level by p4c today?

Reply:
Consulting to my frontend implementation, it seems like `static_assert` is the only case where any kind of call (except constructor calls) is made in the top level.

```p4
struct Headers {
}

struct Meta {
}

const bool test = static_assert(V1MODEL_VERSION >= 20160101, "V1 model version is not >= 20160101");
```

Reply:
So it seems like we might want to define that `static_assert` returns a compile-time known value (I don't see a reason why we need to prepend "local" to that).  The spec already says that the parameters to `static_assert` must be compile-time known values.

Perhaps in the "top level" column, we could say "no" for extern function/method calls, but with a footnote explaining the exception that `static_assert` calls are allowed at the top level?

Reply:
I agree on saying "no" for function/method calls from the top-level, but with a footnote excusing `static_assert`.

Just to clarify, this PR also makes the following change,

```diff
in a select expression. The row for `extern` describes where extern method
- calls can be made from.
+ or function calls can be made from.
```

so extern functions belong to row "extern", and not "function".

And about "local" compile-time known, I suppose it should be "local" because section **11.1 Constants** says:

> The initializer expression must be a local compile-time known value.

, which also suggests that the parameters to `static_assert` should be local compile-time known rather than just compile-time known.

Reply:
If we have an "extern" row and a "function" row, then I believe that (perhaps confusingly, and worth saying explicitly somewhere), that "extern" means "extern function or extern method", and "function" means "function, but NOT and extern function".

It seems reasonable to me to also say "no" for column "top level", row "function", too, for the same reasons that it could get pretty nit-picky very quickly to define which subset of functions are supported.  For example, functions can call extern functions or methods in their body.

I guess that at the top level, everything that is compile-time known is _also_ local compile-time known, because there can be no constructor parameters in scope at the top level.  I don't have a strong preference for whether "lcoal" is there or not.

Reply:
I also think that being compile-time known in the top level would *imply* being a local compile-time known value. But this being the case, I think making it *explicitly* written in the spec (e.g., "`static_assert` returns a local-compile time known value") is more accurate, and also in line with section 18.1 that makes a clear definition of the two concepts.

I guess this is not directly related to this PR, as the revisions to **Appendix F** need not discuss local compile-time known or compile-time known - ness. Maybe there should be a separate PR, revising the section for static_assert :)